app/models/project.rb

```diff
@@ -2,55 +2,118 @@
+  before_save do |project|
+    # Ensure that the metadata JSONB postgres field is persisted properly
+    project.metadata = project.metadata_model
+  end
 
   # Valid project status described in ADR 7
   # See `architecture-decisions/0007-valid-project-statuses.md`
   PENDING_STATUS = "pending"
   APPROVED_STATUS = "approved"
   ACTIVE_STATUS = "active"
 
-  delegate :to_json, to: :metadata_json
+  delegate :to_json, to: :metadata_json # field in the database
+
+  def create!(initial_metadata:, user:)
+    self.metadata_model = initial_metadata
+    if initial_metadata.project_id.blank?
+        self.draft_doi(user: user)
+    end
+    if self.valid?
+      self.save!
+      ProvenanceEvent.generate_submission_events(project: self, user: user)
+
+      #return doi
+      self.metadata_model.project_id
+    else
+      nil
+    end
+  end
+
+  def approve!(mediaflux_id:, current_user:)
+    self.mediaflux_id = mediaflux_id
+    self.metadata_model.status = Project::APPROVED_STATUS
+    self.save!
+
+    # create two provenance events, one for approving the project and
+      # another for changing the status of the project
+    ProvenanceEvent.generate_approval_events(project: self, user: current_user)
+
+  end
+
+  def activate!(collection_id:, current_user:)
+    response = Mediaflux::Http::AssetMetadataRequest.new(session_token: current_user.mediaflux_session, id: collection_id)
+    mediaflux_metadata = response.metadata # get the metadata of the collection from mediaflux
+
+    return unless mediaflux_metadata[:collection] == true # If the collection id exists
+
+    # check if the project doi in rails matches the project doi in mediaflux
+    return unless mediaflux_metadata[:project_id] == self.metadata_model.project_id
 
+    # activate a project by setting the status to 'active'
+    self.metadata_model.status = Project::ACTIVE_STATUS
+
+    # also read in the actual project directory
+    self.metadata_model.project_directory = mediaflux_metadata[:project_directory]
+    self.save!
+
+    ProvenanceEvent.generate_active_events(project: self, user: current_user)
+  end
+
+  def draft_doi(user:)
+    puldatacite = PULDatacite.new
+    self.metadata_model.project_id = puldatacite.draft_doi
+    # self.save!
+  end
+
+  # Ideally this method should return a ProjectMetadata object (like `metadata_model` does)
+  # but we'll keep them both while we are refactoring the code so that we don't break
+  # everything at once since `metadata` is used everywhere.
   def metadata
-    metadata_with_defaults.with_indifferent_access
+    @metadata_hash = (metadata_json || {}).with_indifferent_access
   end
 
   def metadata_model
-    @metadata_model ||= ProjectMetadata.new(project: self)
+    @metadata_model ||= ProjectMetadata.new_from_hash(self.metadata)
+  end
+
+  def metadata_model=(new_metadata_model)
+    @metadata_model = new_metadata_model
   end
 
-  def metadata=(metadata)
-    metadata[:schema_version] ||= TigerdataSchema::SCHEMA_VERSION
```

Comment:
Where is this PR does the current Schema version get defaulted?

Reply:
Done.